### ISSUE TITLE
Support chaining comparison operators in GDScript

### ISSUE THREAD

**Issue by MewPurPur** [godotengine/godot-proposals#3595]
### Describe the project you are working on

Writing code in GDScript

### Describe the problem or limitation you are having in your project

Oftentimes I need two or more comparisons in my conditions, which makes them less readable and less compact.
```
if my_var1 == my_var2 and my_var1 == my_var3 and my_var1 == my_var4
```
Comparisons like this can't be shortened out of the box. Having to repeat the thing twice gets annoying with long variable names, and really annoying with statements like `my_raycast.get_collision_point() - global_position.x`
Another two examples:
```
if my_var > 0 and my_var < 4
if 0 < my_var and my_var < my_other_var
```
You can for example turn the second one into one statement with clamp:
```
if my_var == clamp(my_var, 0, my_other_var)
```
But this doesn't help with readability or compactness.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Chained comparisons would be able to shorten these conditions. For example:
```
if my_var1 == my_var2 == my_var3 == my_var4
```
This will only return true if these variables are all equal.
Only adjacent pairs are checked, so theoretically `1 != 2 != 1` would return true.
Checking if my_var is in range [-1, 1] would look like so: `if -1 <= my_var <= 1`
An added benefit is that you can chain a function to only call it once, without needing a variable for it:
```
if -0.5 < noise(x) < 0.5
```
This can be useful if you don't want to call an expensive function multiple times.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

One way to go about this would be to not even require the same types of comparison operators, and just have the pairwise evaluation. In this case something like `x > y != z <= w` would mean `x > y and y != z and z <= w`.

Another way would be to require comparison operators to be identical. An error would be thrown if a chain of comparisons has different operators.

Alternatively, we could even choose what types of operators can be chained together, or if an operator is even allowed to participate in chaining. For example, allowing `<=` and `<` together would allow chained comparisons to cleanly cover a range like [0, 1). If the operators in a chained comparison are incompatible, an error message would be thrown along the lines of "Incorrect use of `>` and `is` in the same chained comparison".

This feature would also be compatibility-breaking, as individual comparisons in a chain would no longer be treated as bools. To get this functionality back, they would need to be surrounded by brackets (which is a good practice anyway)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I think it would be used often.

### Is there a reason why this should be core and not an add-on in the asset library?

It's a GDScript feature proposal.

**Comment by bluenote10** [comment]
I guess this proposal only makes sense for **inequalities**, chaining `==` is pretty much pointless, right?

Some prior art: Python has this feature, and while nice in some cases, it can also be a source of bugs when operators are overloaded. I.e. for scalar types `lower_bound < x < upper_bound` works, however many Python developers expect that this pattern extends to vectorized expressions using e.g. numpy arrays, but it doesn't.

TL;DR if GDScript ever wants to support comparison operator overloading or vectorized comparison operators (both unlikely?) that feature can be counter productive, otherwise fine :wink:

**Comment by KoBeWi** [comment]
Related to #889 (but more broad)

**Comment by dalexeev** [comment]
I support this proposal. But please note that any comparison (both binary and string) must always return a boolean result, not a numeric one.

> I guess this proposal mainly makes sense for **inequalities**, chaining `==` is a bit unusual, right?

I think all comparison operators should work the same for consistency.

> it can also be a source of bugs when operators are overloaded

Operator overloading is a rather controversial feature in general and in dynamically typed languages especially. Overloaded operators must not violate the original operator semantics. In particular, comparing (more or less) a scalar with a vector is a priori strange.

In addition, the same operator in the context of different types can denote different operations (for example, `+` for numbers is addition, and for strings it is concatenation). But this does not apply to comparison operators.
